tests/firmware_tests.py

```diff
@@ -0,0 +1,346 @@
+
+        sleep_called = any(args and (args[0] == 1 or args[0] == 1.0) for args, kwargs in mock_sleep.call_args_list)
+        self.assertTrue(sleep_called, f"sleep(1) was not called, calls: {mock_sleep.call_args_list}")
+
+        self.assertEqual(client._TBDeviceMqttClient__firmware_request_id, 1)
+        self.assertEqual(client._TBDeviceMqttClient__target_firmware_length, 900)
+        self.assertEqual(client._TBDeviceMqttClient__chunk_count, ceil(900 / 128))
+        client._TBDeviceMqttClient__get_firmware.assert_called_once()
+
+
+class TestTBDeviceMqttClientOnConnect(unittest.TestCase):
+    @patch('tb_device_mqtt.log')
+    def test_on_connect_success(self, mock_logger):
+        client = TBDeviceMqttClient("<THINGSBOARD_HOST>", 1883, "<ACCESS_TOKEN>")
+        client._subscribe_to_topic = MagicMock()
+
+        client._on_connect(client=None, userdata=None, flags=None, result_code=0)
+
+        self.assertTrue(client._TBDeviceMqttClient__is_connected)
+        mock_logger.error.assert_not_called()
+
+        expected_sub_calls = [
+            call('v1/devices/me/attributes', qos=client.quality_of_service),
+            call('v1/devices/me/attributes/response/+', qos=client.quality_of_service),
+            call('v1/devices/me/rpc/request/+', qos=client.quality_of_service),
+            call('v1/devices/me/rpc/response/+', qos=client.quality_of_service),
+        ]
+        client._subscribe_to_topic.assert_has_calls(expected_sub_calls, any_order=False)
+
+        self.assertTrue(client._TBDeviceMqttClient__request_service_configuration_required)
+
+    @patch('tb_device_mqtt.log')
+    def test_on_connect_fail_known_code(self, mock_logger):
+        client = TBDeviceMqttClient("<THINGSBOARD_HOST>", 1883, "<ACCESS_TOKEN>")
+
+        client._on_connect(client=None, userdata=None, flags=None, result_code=1)
+
+        self.assertFalse(client._TBDeviceMqttClient__is_connected)
+        mock_logger.error.assert_called_once_with(
+            "connection FAIL with error %s %s",
+            1,
+            RESULT_CODES[1]
+        )
+
+    @patch('tb_device_mqtt.log')
+    def test_on_connect_fail_unknown_code(self, mock_logger):
+        client = TBDeviceMqttClient("<THINGSBOARD_HOST>", 1883, "<ACCESS_TOKEN>")
+
+        client._on_connect(client=None, userdata=None, flags=None, result_code=999)
+
+        self.assertFalse(client._TBDeviceMqttClient__is_connected)
+        mock_logger.error.assert_called_once_with("connection FAIL with unknown error")
+
+    @patch('tb_device_mqtt.log')
+    def test_on_connect_fail_reasoncodes(self, mock_logger):
+        client = TBDeviceMqttClient("<THINGSBOARD_HOST>", 1883, "<ACCESS_TOKEN>")
+
+        mock_rc = MagicMock(spec=ReasonCodes)
+        mock_rc.getName.return_value = "SomeError"
+
+        client._on_connect(client=None, userdata=None, flags=None, result_code=mock_rc)
+
+        self.assertFalse(client._TBDeviceMqttClient__is_connected)
+        mock_logger.error.assert_called_once_with(
```

Comment:
Using a logger message as a result of the test is not correct.
A test can be affected by simply changing the logger message but tested functionality will still be correct.